.github/files/list-changed-projects.php

```diff
@@ -0,0 +1,273 @@
+	'.github/files/list-changed-projects.php',
+	'.github/files/generate-ci-matrix.php',
+	'.github/files/process-coverage.sh',
+	'.github/files/setup-wordpress-env.sh',
+	'.github/php-version',
+	'.github/workflows/tests.yml',
+	'.nvmrc',
+);
+
+// Globs that should be ignored. Supports `*`, `**`, and `?`.
+// phpcs:ignore VariableAnalysis.CodeAnalysis.VariableAnalysis.UnusedVariable
+$ignore_globs = array(
+	'**.md',
+);
+
+chdir( __DIR__ . '/../../' );
+
+if ( array_search( '--debug', $argv, true ) !== false ) {
+	/**
+	 * Output debug info.
+	 *
+	 * @param array ...$args Arguments to printf. A newline is automatically appended.
+	 */
+	function debug( ...$args ) {
+		$args[0] = "\e[1;30m${args[0]}\e[0m\n";
+		fprintf( STDERR, ...$args );
+	}
+} else {
+	/**
+	 * Do not output debug info.
+	 */
+	function debug() {
+	}
+}
+
+/**
+ * Fetch the list of all projects.
+ *
+ * @return string[]
+ */
+function get_all_projects() {
+	static $cache = null;
+
+	if ( null === $cache ) {
+		$cache = array( 'monorepo' );
+		foreach ( glob( 'projects/*/*/composer.json' ) as $file ) {
+			$cache[] = substr( $file, 9, -14 );
+		}
+		ksort( $cache );
+	}
+
+	return $cache;
+}
+
+/**
+ * Fetch the list of changed projects.
+ *
+ * @return string[]
+ */
+function get_changed_projects() {
+	global $infrastructure_files, $ignore_globs;
+
+	// Code coverage probably needs to run all tests to avoid confusing metrics.
+	// @todo Determine if that's actually true.
+	if ( 'test-coverage' === getenv( 'TEST_SCRIPT' ) ) {
+		debug( 'TEST_SCRIPT is test-coverage, considering all projects changed.' );
+		return get_all_projects();
+	}
+
+	$event = getenv( 'GITHUB_EVENT_NAME' );
+	if ( 'pull_request' === $event ) {
+		$event_path = getenv( 'GITHUB_EVENT_PATH' );
+		if ( ! $event_path || ! file_exists( $event_path ) ) {
+			fprintf( STDERR, "Missing GITHUB_EVENT_PATH for pull_request event\n" );
+			exit( 1 );
+		}
+		$event = json_decode( file_get_contents( $event_path ) );
+		if ( ! isset( $event->pull_request->base->sha ) || ! isset( $event->pull_request->head->sha ) ) {
+			fprintf( STDERR, "Missing pull_request data in GITHUB_EVENT_PATH file\n" );
+			exit( 1 );
+		}
+		$base = $event->pull_request->base->sha;
+		$head = $event->pull_request->head->sha;
+		debug( 'GITHUB_EVENT_NAME is pull_request, checking diff from %s...%s.', $base, $head );
+	} elseif ( 'push' === $event ) {
+		debug( 'GITHUB_EVENT_NAME is push, considering all projects changed.' );
+		return get_all_projects();
+	} else {
+		fprintf( STDERR, "Unsupported GITHUB_EVENT_NAME \"%s\"\n", $event );
+		exit( 1 );
+	}
+
+	$pipes = null;
+	$p     = proc_open(
+		sprintf( 'git diff --no-renames --name-only %s...%s', escapeshellarg( $base ), escapeshellarg( $head ) ),
+		array( array( 'pipe', 'r' ), array( 'pipe', 'w' ), STDERR ),
+		$pipes
+	);
+	if ( ! $p ) {
+		exit( 1 );
+	}
+	fclose( $pipes[0] );
+
+	// PHP doesn't seem to have something that'll correctly handle `**`, so munge to a regex.
```

Comment:
We should add some generic unit tests for monorepo tools. This is one of those things that I would feel better having an unit test to confirm it is handling the `**` as we'd expect.

Could also make a handy small package of itself or general handy php utils.

Reply:
Tests probably would be useful. We can add such tests easily enough now by defining or adding to the appropriate scripts in the root-level `composer.json`, someone would just have to pick a framework and such and then write them.

As its own package, that'd be cool but I'd want to add support for more advanced globs and such.